components/blitz/src/omero/gateway/facility/ROIFacility.java

```diff
@@ -834,71 +846,193 @@ else if (shape.isDirty()) {
+                }
 
-            // get the ROIs of the specified folder
-            IQueryPrx qs = gateway.getQueryService(ctx);
-            StringBuilder sb = new StringBuilder();
-            ParametersI param = new ParametersI();
-            param.addLong("folderId", folderId);
+                IQueryPrx service = gateway.getQueryService(ctx);
+
+                List<IObject> objs = service.findAllByQuery(query, p);
+
+                for (IObject obj : objs)
+                    roiIDs.add(((Roi) obj).getId().getValue());
+            } else if (folderIds.isEmpty()) {
+                // ROIs must not be part of folders
+
+                String query;
+                ParametersI p = new ParametersI();
+                if (imageId == null) {
+                    // ignore imageId, get all rois
+                    query = "select distinct roi from Roi roi "
+                            + "left outer join fetch roi.folderLinks as links ";
+                } else if (imageId > -1) {
+                    // get rois for the image
+                    query = "select distinct roi from Roi roi "
+                            + "left outer join fetch roi.image as image "
+                            + "left outer join fetch roi.folderLinks as links "
+                            + "where image.id = :id";
+                    p.addId(imageId);
+
+                } else {
+                    // get rois not attached to any image
+                    query = "select distinct roi from Roi roi "
+                            + "left outer join fetch roi.folderLinks as links "
+                            + "where roi.image is null";
+                }
 
-            sb.append("select roilink from FolderRoiLink as roilink ");
-            sb.append("left outer join fetch roilink.parent as folder ");
-            sb.append("left outer join fetch roilink.child as roi ");
-            sb.append("where folder.id = :folderId ");
+                IQueryPrx service = gateway.getQueryService(ctx);
 
-            List<IObject> links = qs.findAllByQuery(sb.toString(), param);
+                List<IObject> objs = service.findAllByQuery(query, p);
 
-            Set<Long> roiIds = new HashSet<Long>();
-            for (IObject l : links) {
-                FolderRoiLink link = (FolderRoiLink) l;
-                roiIds.add(link.getChild().getId().getValue());
+                for (IObject obj : objs) {
+                    Roi roi = (Roi) obj;
+                    if (roi.copyFolderLinks().isEmpty())
+                        roiIDs.add(roi.getId().getValue());
+                }
+            } else {
+                // get all rois which are part of the given folders
+                Map<String, Collection<Long>> target = new HashMap<String, Collection<Long>>();
+                target.put("Folder", folderIds);
+
+                FindChildren finder = Requests.findChildren().target(target)
+                        .childType("Roi").build();
```

Comment:
Do we want to include ROIs that are on images that are in the folder?

Also if preferred can simply set target with `.target("Folder").id(folderIds)`